crates/engine/tree/src/tree/mod.rs

```diff
@@ -927,28 +939,31 @@ where
     ///   extension of the canonical chain.
     /// * walking back from the current head to verify that the target hash is not already part of
     ///   the canonical chain.
-    fn is_fork(&self, target_hash: B256, finalized_hash: Option<B256>) -> ProviderResult<bool> {
+    fn is_fork(&self, target_hash: B256, _finalized_hash: Option<B256>) -> ProviderResult<bool> {
         // verify that the given hash is not part of an extension of the canon chain.
+        let canonical_head = self.state.tree_state.canonical_head();
         let mut current_hash = target_hash;
         while let Some(current_block) = self.sealed_header_by_hash(current_hash)? {
-            if current_block.hash() == self.state.tree_state.canonical_block_hash() {
+            if current_block.hash() == canonical_head.hash {
                 return Ok(false)
             }
+            // We already passed the canonical head
+            if current_block.number <= canonical_head.number {
+                break
+            }
```

Comment:
this loop is now safe and cuts off at the canonical head which is correct because this is the section we need to check here.